src/wasm/syscalls_openmp.cpp

```diff
@@ -1,40 +1,109 @@
 #include "WasmModule.h"
-#include "syscalls.h"
-#include "openmp_types.h"
 
 #include <WAVM/Runtime/Runtime.h>
 #include <WAVM/Runtime/Intrinsics.h>
 #include <util/environment.h>
 
+#include <condition_variable>
+#include <mutex>
+#include <stdio.h>
+
 namespace wasm {
     static thread_local int thisThreadNumber = 0;
-    static thread_local unsigned int thisSectionThreadCount = 0;
+    static thread_local unsigned int thisSectionThreadCount = 1;
+    static int masterNumThread = -1;
+    static std::vector<std::thread> threads;
+    static int numThreadSemaphore = -1;
+    static std::mutex barrierMutex;
+    static std::condition_variable barrierCV;
```

Comment:
Will this break if we have concurrent OpenMP functions running?

Reply:
We need to support having multiple teams of threads executing different OpenMP functions as part of the same Faasm runtime process. I have a feeling they would interfere with each other as it is

Reply:
They absolutely would, the next step is to support teams, which this PR is not doing and do like clang which has a global `__kmp_threads` array indexable with the `gtid`. Supporting teams was not in the scope of this PR yet, as it will first require to fix the `gtid` problems I've seen.

Reply:
Ok great, we also have to think about the multi-tenancy angle, i.e. totally isolating two potentially malicious users from each other within the same Faasm runtime. This will be a stricter form of isolation than between teams that are part of the same application. 

If the `gtid` is used to index resources but can be modified in wasm code, then we need to make sure each user has their own `gtid` space. In other words, a malicious user shouldn't be able to manipulate the `gtid` that gets passed to functions to try and access/ mess up another user's stuff.

Reply:
But as you say, out of scope for this PR so will merge

Reply:
That notion of multi-user is something I'm looking at very closely, it is also why I preferred to re-implement the runtime rather than trying to get the current one to compile to WASM for example because I need to study how to integrate control of it.

There are other important questions here on top of the security concerns that you raised, things like fairness, enforceable SLAs, user-defined device & target programming.